File: gallery/pptx/android/desktop/src/main/kotlin/fi/ranger/pptx/desktop/RecordingSurface.kt
package fi.ranger.pptx.desktop

import fi.ranger.pptx.evg.EvgGradient
import fi.ranger.pptx.evg.EvgShadow
import fi.ranger.pptx.evg.EvgSurface

/**
 * A surface that counts what it was asked to draw, and then draws it.
 *
 * "Every slide produced some ink" is a weak thing to check: a deck of nothing
 * but grey boxes passes it, and so does a painter that has quietly stopped
 * dispatching text. What a coverage run wants to know is which *kinds* of
 * command reached a surface at all — whether the corpus exercises paths and
 * clips and gradients and mirrored pictures, and whether the walk delivers
 * them — so this wraps the real surface and keeps a tally on the way through.
 */
class RecordingSurface(private val inner: EvgSurface) : EvgSurface {

    val calls = HashMap<String, Int>()

    private fun bump(what: String) {
        calls[what] = (calls[what] ?: 0) + 1
    }

    fun count(what: String): Int = calls[what] ?: 0

    override fun save() {
        bump("save"); inner.save()
    }

    override fun restore() {
        bump("restore"); inner.restore()
    }

    override fun clipRect(x: Float, y: Float, w: Float, h: Float) {
        bump("clipRect"); inner.clipRect(x, y, w, h)
    }

    override fun rotate(degrees: Float, px: Float, py: Float) {
        bump("rotate"); inner.rotate(degrees, px, py)
    }

    override fun fillRect(
        x: Float, y: Float, w: Float, h: Float,
        radius: Float, color: Int, gradient: EvgGradient?, shadow: EvgShadow?,
    ) {
        bump("fillRect")
        if (radius > 0.5f) bump("roundedRect")
        if (gradient != null) bump("gradient")
        if (shadow != null) bump("shadow")
        inner.fillRect(x, y, w, h, radius, color, gradient, shadow)
    }

    override fun strokeRect(
        x: Float, y: Float, w: Float, h: Float,
        radius: Float, thickness: Float, color: Int,
    ) {
        bump("strokeRect"); inner.strokeRect(x, y, w, h, radius, thickness, color)
    }

    override fun fillPath(
        pts: List<Double>, ringEnds: List<Int>, evenOdd: Boolean,
        color: Int, shadow: EvgShadow?,
    ) {
        bump("fillPath")
        if (ringEnds.size > 1) bump("multiRingPath")
        if (evenOdd) bump("evenOddPath")
        if (shadow != null) bump("shadow")
        inner.fillPath(pts, ringEnds, evenOdd, color, shadow)
    }

    override fun strokePath(pts: List<Double>, ringEnds: List<Int>, thickness: Float, color: Int) {
        bump("strokePath"); inner.strokePath(pts, ringEnds, thickness, color)
    }

    override fun drawImage(
        src: String, x: Float, y: Float, w: Float, h: Float,
        radius: Float, flipH: Boolean, flipV: Boolean,
    ) {
        bump("drawImage")
        if (flipH || flipV) bump("mirroredImage")
        inner.drawImage(src, x, y, w, h, radius, flipH, flipV)
    }

    override fun drawTextRun(
        text: String, x: Float, top: Float, sizePx: Float,
        family: String, bold: Boolean, italic: Boolean, color: Int,
    ) {
        bump("drawTextRun")
        if (bold) bump("boldText")
        if (italic) bump("italicText")
        // A run outside the Latin block is what tells you the TrueType reader
        // and the face pool are doing something rather than getting lucky.
        if (text.any { it.code > 127 }) bump("nonAsciiText")
        inner.drawTextRun(text, x, top, sizePx, family, bold, italic, color)
    }
}

File: gallery/evg/apple/Sources/RecordingSurface.swift
// SPDX-License-Identifier: AGPL-3.0-or-later
//
// A surface that draws nothing and counts everything.
//
// Wrap a real surface in one of these and the frame is painted exactly as it
// would be, with a tally of what was dispatched on the other side. "Some ink
// appeared" is a weak check — a painter that had quietly stopped drawing text
// would pass it — so a check asserts the counts instead: this page reaches
// text, filled boxes, borders, rounded corners, clipping, a filled path and
// stroked paths, and every `save` is matched by a `restore`.
//
// The Ranger side of this port counts the same things one step earlier, on the
// display list itself (`gallery/ui/ios/ranger/check_ios.rgr`), because that can
// run without a Mac. This is for the half that cannot: it says whether the
// PAINTER dispatched them, not merely whether the page produced them.

import CoreGraphics

public final class RecordingSurface: EvgSurface {

    public private(set) var calls: [String: Int] = [:]
    private let inner: EvgSurface?

    public init(wrapping inner: EvgSurface? = nil) {
        self.inner = inner
    }

    public func count(_ kind: String) -> Int {
        calls[kind] ?? 0
    }

    private func note(_ kind: String) {
        calls[kind, default: 0] += 1
    }

    public var summary: String {
        calls.keys.sorted().map { "\($0)=\(calls[$0]!)" }.joined(separator: "  ")
    }

    public func save() {
        note("save")
        inner?.save()
    }

    public func restore() {
        note("restore")
        inner?.restore()
    }

    public func clipRect(x: CGFloat, y: CGFloat, w: CGFloat, h: CGFloat) {
        note("clipRect")
        inner?.clipRect(x: x, y: y, w: w, h: h)
    }

    public func rotate(degrees: CGFloat, px: CGFloat, py: CGFloat) {
        note("rotate")
        inner?.rotate(degrees: degrees, px: px, py: py)
    }

    public func fillRect(
        x: CGFloat, y: CGFloat, w: CGFloat, h: CGFloat,
        radius: CGFloat, color: EvgColor,
        gradient: EvgGradient?, shadow: EvgShadow?
    ) {
        note("fillRect")
        if radius > 0 { note("roundedRect") }
        if gradient != nil { note("gradient") }
        if shadow != nil { note("shadow") }
        inner?.fillRect(x: x, y: y, w: w, h: h, radius: radius, color: color, gradient: gradient, shadow: shadow)
    }

    public func strokeRect(
        x: CGFloat, y: CGFloat, w: CGFloat, h: CGFloat,
        radius: CGFloat, thickness: CGFloat, color: EvgColor
    ) {
        note("strokeRect")
        inner?.strokeRect(x: x, y: y, w: w, h: h, radius: radius, thickness: thickness, color: color)
    }

    public func fillPath(
        pts: [Double], ringEnds: [Int], evenOdd: Bool,
        color: EvgColor, shadow: EvgShadow?
    ) {
        note("fillPath")
        if shadow != nil { note("shadow") }
        inner?.fillPath(pts: pts, ringEnds: ringEnds, evenOdd: evenOdd, color: color, shadow: shadow)
    }

    public func strokePath(pts: [Double], ringEnds: [Int], thickness: CGFloat, color: EvgColor) {
        note("strokePath")
        inner?.strokePath(pts: pts, ringEnds: ringEnds, thickness: thickness, color: color)
    }

    public func drawImage(
        src: String, x: CGFloat, y: CGFloat, w: CGFloat, h: CGFloat,
        radius: CGFloat, flipH: Bool, flipV: Bool
    ) {
        note("drawImage")
        inner?.drawImage(src: src, x: x, y: y, w: w, h: h, radius: radius, flipH: flipH, flipV: flipV)
    }

    public func drawTextRun(
        text: String, x: CGFloat, top: CGFloat, sizePx: CGFloat,
        family: String, bold: Bool, italic: Bool, color: EvgColor
    ) {
        note("drawTextRun")
        if italic { note("italicText") }
        inner?.drawTextRun(text: text, x: x, top: top, sizePx: sizePx, family: family, bold: bold, italic: italic, color: color)
    }
}
